68276 new purge api

jiangphcn opened the pull request.

davisp commented:
LGTM though we can't merge this until after PSE stuff lands of course.

I just have a few minor style tweaks to note but otherwise this looks pretty good all around.

davisp commented on src/couch_mrview_cleanup.erl:
```diff
@@ -41,7 +41,20 @@ run(Db) ->
+        % delete local purge document if have
+        [PureFN] = lists:nthtail(length(FilePathList)-1, FilePathList),
+        PureFNExt = filename:extension(PureFN),
+        Sig = filename:basename(PureFN, PureFNExt),
```
You should take the last four or five lines and move them to a couch_mrview_util function called get_signature_from_filename or something similar. We should also probably assert that the signature is something close to expected, ie > 16 characters with only hexadecimal characters or something.

davisp commented on src/couch_mrview_index.erl:
```diff
@@ -204,3 +208,53 @@ index_file_exists(State) ->
+            false
+    end.
+
+
```
You've got two extra blank lines here.

davisp commented on src/couch_mrview_util.erl:
```diff
@@ -38,6 +39,26 @@
 -include_lib("couch/include/couch_db.hrl").
 -include_lib("couch_mrview/include/couch_mrview.hrl").
 
+utc_string() ->
```
Missing a blank line before utc_string/0

davisp commented on src/couch_mrview_index.erl:
```diff
@@ -204,3 +208,53 @@ index_file_exists(State) ->
+            {<<"indexname">>, get(idx_name, State)},
+            {<<"signature">>, Sig}
+        ]}},
+        {<<"type">>, <<"view">>}
```
I would use <<"mrview">> instead of just <<"view">>.

davisp commented on src/couch_mrview_util.erl:
```diff
@@ -38,6 +39,26 @@
+
+
+get_local_purge_doc_id(Sig) ->
+    list_to_binary(?LOCAL_DOC_PREFIX ++ "purge-view-" ++ Sig).
```
I would use purge-mrview instead of purge-view

jiangphcn commented:
Thanks, Paul. I already addressed all comments above in commit 4a1d645. Please help check when you get time. Regarding `utc_string()`, we plan to ask Mayya to create one in couchdb-couch and refer to new one in other places.

davisp commented on src/couch_mrview_index.erl:
```diff
@@ -204,3 +208,56 @@ index_file_exists(State) ->
+    case couch_db:open_int(ShardDbName, []) of
+        {ok, Db} ->
+            try
+                DbName =mem3:dbname(Db#db.name),
```
Missing space after = operator.

davisp commented on src/couch_mrview_index.erl:
```diff
@@ -204,3 +208,56 @@ index_file_exists(State) ->
+
+maybe_create_local_purge_doc(Db, State) ->
+    Sig = couch_index_util:hexsig(get(signature, State)),
+    case couch_db:open_doc(Db, couch_mrview_util:get_local_purge_doc_id(Sig), []) of
```
Line is longer than 80 characters.

davisp commented on src/couch_mrview_util.erl:
```diff
@@ -39,6 +41,26 @@
+
+get_signature_from_filename(FileName) ->
+    FilePathList = filename:split(FileName),
+    [PureFN] = lists:nthtail(length(FilePathList)-1, FilePathList),
```
Missing spaces around - operator.

davisp commented on test/couch_mrview_purge_docs_tests.erl:
```diff
@@ -0,0 +1,126 @@
+        ]},
+        ?assertEqual(Expect, Result),
+
+        % 1st purge request
```
Unnecessary comment

davisp commented on src/couch_mrview_index.erl:
```diff
@@ -204,3 +208,56 @@ index_file_exists(State) ->
+        {<<"verify_function">>, <<"verify_index_exists">>},
+        {<<"verify_options">>, {[
+            {<<"dbname">>, get(db_name, State)},
+            {<<"indexname">>, get(idx_name, State)},
```
you should call this ddoc_id here since that'll be more consistent with other parts of the code base. For other indexers that don't treat the whole ddoc as a single index is where there's an added indexname.

davisp commented on src/couch_mrview_index.erl:
```diff
@@ -230,20 +230,23 @@ update_local_purge_doc(Db, State) ->
+            catch E:T ->
+                Stack = erlang:get_stacktrace(),
+                couch_log:error("Error occurs when verifying existence of index: ~p ~p",
+                    [{E, T}, Stack]),
```
I would change that to ... "verifying existince of ~s/~s :: ~p ~p", [ShardDbName, DDocId, {E, T}, Stack]
